consensus/qsc/history.go

```diff
@@ -0,0 +1,198 @@
+package qsc
+
+import (
+	"bytes"
+	fmt "fmt"
+	"strings"
+
+	proto "github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes"
+	"go.dedis.ch/fabric/encoding"
+)
+
+// epoch is a part of an history. In other words, it represents a time step in
+// the QSC algorithm.
+//
+// - implements encoding.Packable
+type epoch struct {
+	hash   []byte
+	random int64
+}
+
+// Pack implements encoding.Packable. It returns the protobuf message for an
+// epoch.
+func (e epoch) Pack() (proto.Message, error) {
+	pb := &Epoch{
+		Random: e.random,
+		Hash:   e.hash,
+	}
+
+	return pb, nil
+}
+
+// Equal returns true when the other epoch is the same as the current one.
+func (e epoch) Equal(other epoch) bool {
+	if e.random != other.random {
+		return false
+	}
+	if !bytes.Equal(e.hash, other.hash) {
+		return false
+	}
+
+	return true
+}
+
+// history is an ordered list of epochs, or a list of QSC time steps.
+//
+// - implements encoding.Packable
+// - implements fmt.Stringer
+type history []epoch
+
+func (h history) getLast() (epoch, bool) {
+	if len(h) == 0 {
+		return epoch{}, false
+	}
+
+	return h[len(h)-1], true
+}
+
+// Equal returns true when both histories are equal, false otherwise.
+func (h history) Equal(other history) bool {
+	if len(h) != len(other) {
+		return false
+	}
+
+	for i, e := range h {
+		if !e.Equal(other[i]) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// Pack implements encoding.Packable. It returns the protobuf message for an
+// history.
+func (h history) Pack() (proto.Message, error) {
+	pb := &History{
+		Epochs: make([]*Epoch, len(h)),
+	}
+
+	for i, epoch := range h {
+		packed, err := epoch.Pack()
+		if err != nil {
+			return nil, err
+		}
+
+		pb.Epochs[i] = packed.(*Epoch)
+	}
+
+	return pb, nil
+}
+
+// String implements fmt.Stringer. It returns a string representation of the
+// history.
+func (h history) String() string {
+	epochs := make([]string, len(h))
+	for i, e := range h {
+		if len(e.hash) >= 2 {
+			epochs[i] = fmt.Sprintf("%x", e.hash)[:4]
+		} else {
+			epochs[i] = "nil"
+		}
+	}
+	return fmt.Sprintf("History[%d]{%s}", len(h), strings.Join(epochs, ","))
```

Comment:
I've never been a big fan of that because you limit the representation structure. In my opinion, the String function should return a one-line representation, and if you need a detailed description, you iterate through the fields and create your own representation. But that works only for accessible fields..